pkg/kubelet/kubelet_node_status.go

```diff
@@ -553,3 +570,57 @@ func validateNodeIP(nodeIP net.IP) error {
 	}
 	return fmt.Errorf("Node IP: %q not found in the host's network interfaces", nodeIP.String())
 }
+
+// nodeStatusHasChanged compares the original node and current node's status and
+// returns true if any change happens. The heartbeat timestamp is ignored.
+func nodeStatusHasChanged(originalStatus *v1.NodeStatus, status *v1.NodeStatus) bool {
+	if originalStatus == nil && status == nil {
+		return false
+	}
+	if originalStatus == nil || status == nil {
+		return true
+	}
+
+	if nodeConditionsHaveChanged(originalStatus.Conditions, status.Conditions) {
+		return true
+	}
+
+	// Setting the conditions to nil to compare the rest of the node status.
+	// Restoring the conditions in the defer block before returns
+	originalConditions := originalStatus.Conditions
+	originalStatus.Conditions = nil
+	conditions := status.Conditions
+	status.Conditions = nil
+	defer func() {
+		originalStatus.Conditions = originalConditions
+		status.Conditions = conditions
+	}()
+	return !apiequality.Semantic.DeepEqual(originalStatus, status)
+}
+
+// nodeConditionsHaveChanged compares the original node and current node's
+// conditions and returns true if any change happens. The heartbeat timestamp is
+// ignored.
+func nodeConditionsHaveChanged(originalConditions []v1.NodeCondition, conditions []v1.NodeCondition) bool {
+	if len(originalConditions) != len(conditions) {
+		return true
+	}
+
+	originalConditionsCopy := make([]v1.NodeCondition, 0, len(originalConditions))
+	originalConditionsCopy = append(originalConditionsCopy, originalConditions...)
+	conditionsCopy := make([]v1.NodeCondition, 0, len(conditions))
+	conditionsCopy = append(conditionsCopy, conditions...)
+
+	sort.Slice(originalConditionsCopy, func(i, j int) bool { return originalConditionsCopy[i].Type < originalConditionsCopy[j].Type })
```

Comment:
The conditions are merged by keys, where the keys are the type: "patchStrategy:"merge" patchMergeKey:"type"". See https://github.com/kubernetes/kubernetes/blob/8dcdec0a6700b04beb9f73b66f5b125297c9acc7/staging/src/k8s.io/api/core/v1/types.go#L3962

Reply:
I see those, but API server validation does not seem to enforce that uniqueness

Reply:
In reality, both node status setter and node lifecycle controller check if node condition preexists and would just reuse if it is. So I would argue it is enough to just rely on types for sorting here.

Are you suggesting we compare both NodeConditionType and ConditionStatus during sorting?

Reply:
I just changed `sort.Slice` to `sort.SliceStable`, so that the order of equal values are preserved.

Since kubelet node status setters check if the condition preexists, kubelet does not change the order of the conditions. Then using `sort.SliceStable` means the sorted order is deterministic.